File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:

    void findnextgreatestelement(vector<int>& arr, int target){
        int st=0,en=arr.size();
        int mid;
        while(st<en){
            mid = (st+en)/2;
            if(arr[mid]==target){return;}
            else if(arr[mid]<target){
                st = mid+1;
            }else{
                en = mid;
            }
        }
        arr[st] = target;return;
    }


    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> v;
        v.push_back(nums[0]);
        
        for(int i=1;i<n;i++){
            if(nums[i]>v.back()){
              v.push_back(nums[i]);  
            }else{
              findnextgreatestelement(v, nums[i]);
            }
        }
        return v.size();
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {

    public void findnextgreaterelement(List<Integer> nums,int target){
        int st = 0,en=nums.size();
        int mid;

        while(st<en){
            mid = (st+en)/2;
            int num = nums.get(mid);
            if(num==target){
                return;
            }else if(num<target){
                st = mid+1;
            }else{
                en = mid;
            }
        }
        nums.set(st,target);
        return;
    }

    public int lengthOfLIS(int[] nums) {
        List<Integer> st = new ArrayList<>();
        st.add(nums[0]);
        int ans = 1;
        for(int i=1;i<nums.length;i++){
            if(nums[i]>st.get(st.size()-1)){
                st.add(nums[i]);
            }else{
                findnextgreaterelement(st,nums[i]);
            }
        }
        return st.size();
    }
}